src/jdk.compiler/share/classes/com/sun/tools/javac/code/Types.java

```diff
@@ -2236,60 +2235,60 @@ public Type visitErrorType(ErrorType t, Symbol sym) {
+     * sequence.
      *
      * @param t a type
      * @param sym a symbol
      */
     public Type asOuterSuper(Type t, Symbol sym) {
-        switch (t.getTag()) {
-        case CLASS:
-            do {
-                Type s = asSuper(t, sym);
-                if (s != null) return s;
-                t = t.getEnclosingType();
-            } while (t.hasTag(CLASS));
-            return null;
-        case ARRAY:
-            return isSubtype(t, sym.type) ? sym.type : null;
-        case TYPEVAR:
-            return asSuper(t, sym);
-        case ERROR:
-            return t;
-        default:
-            return null;
-        }
+        return asSuperClosure(t, sym, Type::getEnclosingType);
     }
 
     /**
-     * Return the base type of t or any of its enclosing types that
-     * starts with the given symbol.  If none exists, return null.
+     * Traverses a sequence of types starting with `t` and returns the first type
+     * that can be seen as a supertype of one of those types.
+     *
+     * The sequence of types starts with `t` and the next type in the sequence
+     * is obtained by obtaining innermost lexically enclosing class type of the
+     * previous type in the sequence.
      *
      * @param t a type
      * @param sym a symbol
      */
     public Type asEnclosingSuper(Type t, Symbol sym) {
-        switch (t.getTag()) {
-        case CLASS:
-            do {
-                Type s = asSuper(t, sym);
-                if (s != null) return s;
-                Type outer = t.getEnclosingType();
-                t = (outer.hasTag(CLASS)) ? outer :
-                    (t.tsym.owner.enclClass() != null) ? t.tsym.owner.enclClass().type :
-                    Type.noType;
-            } while (t.hasTag(CLASS));
-            return null;
-        case ARRAY:
-            return isSubtype(t, sym.type) ? sym.type : null;
-        case TYPEVAR:
-            return asSuper(t, sym);
-        case ERROR:
-            return t;
-        default:
-            return null;
+        return asSuperClosure(t, sym, type -> getOwnerEnclosingClassType(type));
+    }
+    // where
+    private static Type getOwnerEnclosingClassType(Type type) {
+        return (type.tsym.owner.enclClass() != null)
+                ? type.tsym.owner.enclClass().type
+                : Type.noType;
+    }
+
+    /**
+     * Traverses a sequence of types starting with `t` and returns the first type
+     * that can be seen as a supertype of one of those types. The sequence of types
+     * starts with `t` and is left abstract.
+     *
+     * The next type in the sequence is obtained by passing the previous type in
+     * the sequence to the unary operator `nextType`.
+     *
+     * @param t a type
+     * @param sym a symbol
+     * @param nextType a unary operator that emits the next type to be examined
+     */
+    public Type asSuperClosure(Type t, Symbol sym, UnaryOperator<Type> nextType) {
```

Comment:
Thinking more about naming -- javac is using the term `closure` typically to denote method that return a "bag of stuff" (e.g. transitive _closure_ of supertypes, members, etc.). Here `closure` doesn't mean really that, so I'm having second thoughts :-(

Reply:
It also occurred to me that this `asSuperClosure` we're trying to add can be expressed using streams (I think):

```
Type input = ...
UnaryFunction<Type> updateFunction = ...
Function<Type, Type> mapFunction = ...

Stream.iterate(input, updateFunction)
      .map(mapFunction)
      .filter(site -> !site.hasTag(NONE))
      .findFirst();
```

Where `mapFunction` is `asSuper(t, sym)`, and `updateFunction` differs, depending on whether we want `asEnclosingSuper` or `asOuterSuper`.

Reply:
My other feeling is that, while there is indeed some hidden commonality between `asEnclosingSuper` and `asOuterSuper` -- perhaps the addition of this new method is not super interesting -- as it is effectively only used in these two cases -- and it seems too specific to be used anywhere else (because it always uses `asSuper` and cannot be customized much). So perhaps, for now, the more honest approach is to leave it out.